scripts/get-nightly-build-pass-rate.ts

```diff
@@ -0,0 +1,108 @@
+dotenv.config();
+
+const loginAndAnalyzeBuilds = async (
+  latestBuildNumber: number,
+  numberOfDays: number,
+  username: string,
+  token: string
+) => {
+  const loginUrl = process.env.JENKINS_LOGIN_URL;
+  const baseApiUrl = process.env.JENKINS_BASE_API_URL;
+  const browser = await puppeteer.launch({ headless: true });
+  const page = await browser.newPage();
+
+  if (!loginUrl || !baseApiUrl) {
+    console.error('Missing environment variables. Please check your .env file.');
+    return;
+  }
+
+  try {
+    console.log('Logging in to Jenkins...');
+    await page.goto(loginUrl, { waitUntil: 'networkidle2' });
+    await page.type('#j_username', username);
+    await page.type('input[name="j_password"]', token);
+    await Promise.all([
+      page.click('button[type="submit"]'),
+      page.waitForNavigation({ waitUntil: 'networkidle2' }),
+    ]);
+    console.log('Login successful.');
+
+    const startBuildNumber = latestBuildNumber - numberOfDays + 1;
+    console.log(`Analyzing builds from ${startBuildNumber} to ${latestBuildNumber}...`);
+
+    const failedBuilds: number[] = [];
+    let passCount = 0;
+
+    for (let buildNumber = startBuildNumber; buildNumber <= latestBuildNumber; buildNumber++) {
+      console.log(`Fetching data for build #${buildNumber}...`);
+      const url = `${baseApiUrl}/${buildNumber}/api/json`;
+      try {
+        const response = await page.goto(url, { waitUntil: 'networkidle0', timeout: 10000 });
+
+        if (!response) {
+          console.log(`Build #${buildNumber}: No response from server.`);
+          continue;
+        }
+
+        const responseBody = await response.text();
+
+        if (!responseBody) {
+          console.log(`Build #${buildNumber}: Empty response body.`);
+          continue;
+        }
+
+        let data: { result?: string } | null = null;
+        try {
+          data = JSON.parse(responseBody);
+        } catch (jsonError) {
+          console.error(`Build #${buildNumber}: Failed to parse JSON.`, jsonError);
+          continue;
+        }
+
+        if (!data || !data.result) {
+          console.log(`Build #${buildNumber}: Data unavailable or missing result.`);
+          continue;
+        }
+
+        if (data.result === 'FAILURE') {
+          failedBuilds.push(buildNumber);
+        } else if (data.result === 'SUCCESS') {
+          passCount++;
+        } else {
+          console.log(`Build #${buildNumber}: Unknown result.`);
+        }
+      } catch (error) {
+        console.error(`Error fetching build #${buildNumber}: ${(error as Error).message}`);
+      }
+    }
+
+    const failCount = failedBuilds.length;
+    const passToFailRate = (passCount / (passCount + failCount)) * 100;
+
+    console.log(`Analysis complete.`);
+    console.log(`Pass Count: ${passCount}`);
+    console.log(`Fail Count: ${failCount}`);
+    console.log(`Pass-to-Fail Rate: ${passToFailRate.toFixed(2)}%`);
+    console.log(`Failed Builds: ${failedBuilds.join(', ')}`);
```

Comment:
This is looking great! I would love to have these results written to a file. Do you think you could write these out to a CSV with 4 columns? PI number, pass count, fail count, ratio and maybe a list of the builds that were not accounted for for whatever reason. Does that sound like a good structure/reasonable?

Reply:
Oh yes, good call! I'd like to see better formatting for this output.

Reply:
@jack-ryan-nava-pbc, After giving it some more thought, I think it might be best to hold off on finalizing the output format. I discussed some possible new features with @kim-cmsds earlier, and I foresee that we might be adding some elements down the road that could impact how we structure our data output. So to avoid revisions, I'd like to delay this for now. What do you think?

Reply:
Sounds good! I'm totally onboard.